homeassistant/components/recorder/core.py

```diff
@@ -1021,29 +972,28 @@ def _process_non_state_changed_event_into_session(self, event: Event) -> None:
             event_session.add(dbevent)
             return
 
-        if not (shared_data_bytes := self._serialize_event_data_from_event(event)):
+        event_data_manager = self.event_data_manager
+        if not (shared_data_bytes := event_data_manager.serialize_from_event(event)):
             return
 
         # Map the event data to the EventData table
         shared_data = shared_data_bytes.decode("utf-8")
         # Matching attributes found in the pending commit
-        if pending_event_data := self._pending_event_data.get(shared_data):
+        if pending_event_data := event_data_manager.get_pending(shared_data):
             dbevent.event_data_rel = pending_event_data
         # Matching attributes id found in the cache
-        elif data_id := self._event_data_ids.get(shared_data):
+        elif (data_hash := EventData.hash_shared_data_bytes(shared_data_bytes)) and (
+            data_id := self.event_data_manager.get(
```

Comment:
It would be better to avoid the hash here if we have it in the cache since it can be expensive.